sirepo/template/srw.py

```diff
@@ -636,13 +636,9 @@ def post_execution_processing(
     success_exit=True, is_parallel=True, run_dir=None, **kwargs
 ):
     if success_exit:
-        # Should sim_type and sim_id be available in kwargs?
-        if _SIM_DATA.ML_REPORT in str(run_dir):
+        if _SIM_DATA.is_for_ml(kwargs.get("compute_model")):
```

Comment:
Here define `compute_model` as an argument, since it is a required argument. This code is not fail fast, which it should be. The reason for kwargs is to allow dispatched routines which do not use new parameters to continue to operate when new parameters are added to the interface. This future proofs the dispatchee's from changes to the dispatcher.

Reply:
I presume a named argument (`compute_model=None...assert(compute_model)`) or all the other templates have to change. Also, same treatment for sim_id?

Reply:
You don't need to add compute_model=None when all the other implementations of post_execution_processing have **kwargs. That's what kwargs is for. However, srw.post_execution_processing(success_exit, compute_model, ...., **kwargs) will fail if compute_model is not in kwargs at the dispatch point (in job_cmd). This way you can have a required parameter in some dispatchees that verifies that the dispatcher is supplying a specific result. You should not have:

def post_execution_processing(
    success_exit=True, is_parallel=True, run_dir=None, **kwargs
):

because these are *required* parameters and we want fast fail.

It's important to think through this: what does a dispatcher provide? What does a dispatched call expect? How does this interact with kwargs and named parameters?

Reply:
I did not know you could pull named parameters out of kwargs to use as positionals - I thought they were mutually exclusive. You can't go the other way and _call_ a  function with a positional and kwargs that includes that same name. So this should be
```python
def post_execution_processing(
    compute_model, sim_id, success_exit, **kwargs,
):
```
If that's right, should I change all the other templates, or limit the scope?

Reply:
Yes, remove the defaults as they are incorrect, but you can leave the params as they are used.

Reply:
> I did not know you could pull named parameters out of kwargs to use as positionals - I thought they were mutually exclusive. You can't go the other way and call a function with a positional and kwargs that includes that same name. 

Yes the implementation is correct. I think the language needs to be clear.

The word "positional" is only relevant when talking about [arguments](https://docs.python.org/3/glossary.html#term-argument), that is, the values passed to a call of a function. There are no positional [parameters](https://docs.python.org/3/glossary.html#term-parameter). All parameters have a position in a definition and a name.

The parameter `**<anything>` (commonly called `kwargs`) collects any remaining unnamed *arguments*. And `*<anything>` (commonly `args`) collects all positional arguments that are mapped to named parameters.